docs/architecture/adr-010-monitoring.md

```diff
@@ -0,0 +1,103 @@
+metrics and, maybe later, traces of transactions and RPC queries. See
+https://github.com/tendermint/tendermint/issues/986.
+
+A few solutions were considered:
+
+1. [Prometheus](https://prometheus.io)
+  a) Prometheus API
+  b) [go-kit metrics package](https://github.com/go-kit/kit/tree/master/metrics) as an interface plus Prometheus
+  c) [telegraf](https://github.com/influxdata/telegraf)
+  d) new service, which will listen to events emitted by pubsub and report metrics
+5. [OpenCensus](https://opencensus.io/go/index.html)
+
+### 1. Prometheus
+
+Prometheus seems to be the most popular product out there for monitoring. It has
+a Go client library, powerful queries, alerts.
+
+**a) Prometheus API**
+
+We can commit to using Prometheus in Tendermint, but I think Tendermint users
+should be free to choose whatever monitoring tool they feel will better suit
+their needs (if they don't have existing one already). So we should try to
+abstract interface enough so people can switch between Prometheus and other
+similar tools.
+
+**b) go-kit metrics package as an interface**
+
+metrics package provides a set of uniform interfaces for service
+instrumentation and offers adapters to popular metrics packages:
+
+https://godoc.org/github.com/go-kit/kit/metrics#pkg-subdirectories
+
+Comparing to Prometheus API, we're losing customisability and control, but gaining
+freedom in choosing any instrument from the above list given we will extract
+metrics creation into a separate function (see "providers" in node/node.go).
+
+**c) telegraf**
+
+Unlike already discussed options, telegraf does not require modifying Tendermint
+source code. You create something called an input plugin, which polls
+Tendermint RPC every second and calculates the metrics itself.
+
+While it may sound good, but some metrics we want to report are not exposed via
+RPC or pubsub, therefore can't be accessed externally.
+
+**d) service, listening to pubsub**
+
+Same issue as the above.
+
+### 2. opencensus
+
+opencensus provides both metrics and tracing, which may be important in the
+future. It's API looks different from go-kit and Prometheus, but looks like it
+covers everything we need.
+
+Unfortunately, OpenCensus go client does not define any
+interfaces, so if we want to abstract away metrics we
+will need to write interfaces ourselves.
+
+### List of metrics
+
+|   | Name                                    | Type    |                                                                               |
+| - | --------------------------------------- | ------- | ----------------------------------------------------------------------------- |
+| A | height                                  | Counter |                                                                               |
+| A | validators:<height>                     | Gauge   | Number of validators who signed                                               |
```

Comment:
`<height>` has a couple of occurrences, which I suspect will lead to high cardinality in our metrics, something which can be a problem for long-running processes. Are we going to cap how many of them we keep around, or do we just expose the latest height?

Reply:
I thought of it as a label. Like `validators` with a label `height-11`. Sorry for confusion. What do you think it?